### ISSUE TITLE
[RFC] Support safe embedding

### ISSUE THREAD

**Issue by stephanemagnenat** [RustPython/RustPython#4210]
## Summary

As far as I have seen, RustPython is not yet suitable for "safe" embedding, meaning that executed Python code can block or hurt the caller code, because:

- execution can loop forever,
- memory allocation is not bounded,
- import cannot be disabled.

## Detailed Explanation

I wish to use RustPython as a scripting language within a game engine, running third-party user code. A requirement for me is that this code is run in a safe way. As far as I have seen (but I might have missed some elements), it is currently not the case:

- It is not possible to limit code execution up to a maximum number of instructions. Looking into RustPython's source, if I'm not mistaken, this feature would need to be added to `ExecutingFrame` in its [execution loop](https://github.com/RustPython/RustPython/blob/main/vm/src/frame.rs#L347). Maybe the [`ExecutionResult` type](https://github.com/RustPython/RustPython/blob/main/vm/src/frame.rs#L123) could be extended with an `InstructionBudgetExceeded` variant or similar.
- I couldn't find a way to set limits to heap allocation or stack size, that would also be needed.
- My understanding is that [importlib is always enabled](https://github.com/RustPython/RustPython/blob/main/vm/src/import.rs#L34), but that the OS module is currently disabled in [Wasm32 or Wasi](https://github.com/RustPython/RustPython/blob/main/vm/src/import.rs#L42), is that correct? If so, there should be a way to have finer-control on import lib or even forbid user-defined import statements altogether. Also, I see no reason to link the control of the availability of the `os` module to WASM, maybe a feature flag would be a good addition. Similarly, it should be possible to not enable some [Windows specific code](https://github.com/RustPython/RustPython/blob/main/vm/src/import.rs#L45) and fully disable or controls IOs regardless of the platform (a game engine is a platform in itself).

## Drawbacks, Rationale, and Alternatives

The rationale is to use RustPython as an embedding language within larger software, such as game engines. In these, the software must fully control the scripting environment's limits.

The main drawback is increased code complexity within RustPython, but I believe it can be done cleanly, with some work of course. The split of the Std library (#3102) was already a step towards the direction of embedding.

The alternatives are to not implement this feature, or do it in a fork. A similar issue exists (#3090), but it is more of a question, so I thought an RFC-style new issue is better.

## Unresolved Questions

There are quite some design questions obviously, but I guess first one should agree whether this overall feature makes sense for the project, then the design can be worked out. Probably a unified way to control embedding would be elegant.

**Comment by stephanemagnenat** [comment]
Thank you for your quick answer! Yes, our need is indeed more about sandboxing beyond simple embedding.

Regarding our use case, we are creating an educational game creation platform (https://cand.li) that currently has its custom visual language and is written in a mix of Typescript and Rust (compiled to WASM). We are progressively migrating more and more of our code base to Rust, and in the medium term Python support is something our customers (teachers and students) are asking for, the idea being that they can start with the visual language, then write their own advanced blocks in Python, even sharing them with less-advanced students. However, sandboxing is critical:
- As games can be shared, arbitrary user-written code will be executed on the target machine.
- As users are typically students, it is likely they will sometimes write infinite loops and memory-exploding algorithms, even during development. So to provide the proper educational experiences, these bugs must be detected and reported (I guess some advanced static code analysis could help but is likely not enough).

Regarding memory quota, I imagine that it would be possible for all internal Python data to go through a gate checking the quota when being created, and basically fail execution if the quota is reached (for us that would be enough). Of course I guess it would be an optional feature. I'm under the impression that it touches a similar problematic as the garbage collection, and maybe can be designed at the same time.

**Comment by youknowone** [comment]
The IO parts looks like to be natively blocked by wasm environment.
Yes, we basically are interested in sandboxing, but there wasn't much progress yet.

**Comment by stephanemagnenat** [comment]
> The IO parts looks like to be natively blocked by wasm environment.

Yes, but at least in our case, we would like to have it on native as well, as we plan to make a native app at some point. I imagine that other game projects would have a similar need.

> Yes, we basically are interested in sandboxing, but there wasn't much progress yet.

Good to hear that you are interested! I'm happy to provide input from the use case side of the thing.

**Comment by BjornTheProgrammer** [comment]
Hey, I was just wondering if this is possible in the latest version of RustPython. I wish to run a native app that can run user-generated Python code. Ideally, it would be as feature-rich as possible but without access to IO of any type. An alternative I have thought of is to use a wasm build and call the interpreter through that to do sandboxing.

**Comment by FlippingBinary** [comment]
I'm very interested in using this library in my projects because Python is such a popular language in data science, but stability in the face of arbitrary code is very important to me. I wouldn't want a user to be able to accidentally crash their browser or lock up their machine. 

To limit the amount of memory (or even the rate at which memory is allocated), perhaps the library could support [custom allocators](https://nical.github.io/posts/rust-custom-allocators.html)? Rust uses the `Global` one by default, but I think RustPython could add support for custom allocators without disrupting any existing code that depends on it. Then the user could create a custom allocator that keeps track of whatever matters to them and denies allocation (pretending to be out of memory) when a resource limit has been reached. The simplest approach for the RustPython library would be to simply expose the ability to override the custom allocator for the memory allocated by the vm for the Python code, but another option would be to define one or more custom allocators that have a nice API for setting memory limits.

For execution duration, obviously a multi-threaded application could just use a dedicated thread for the vm and terminate it whenever it wants, but a single-threaded application won't have that luxury. Is there a way to get an iterator from the vm instead of trusting that `run` will eventually return? An iterator could be stepped for as long as the calling code wants until some number of iterations have been reached or the wall clock passes some duration of time. If the iterator could also return the size of the stack (at least the depth of it), the calling code could enforce a limit on how deep the stack can grow, breaking out of iteration when it suspects infinite recursion.

**Comment by NoelJacob** [comment]
> Ideally, it would be as feature-rich as possible but without access to IO of any type. 

Did you find anyway to do it?
> An alternative I have thought of is to use a wasm build and call the interpreter through that to do sandboxing.

Did you try that?

**Comment by BjornTheProgrammer** [comment]
@NoelJacob I did try it, and it worked, however it is complicated. I cannot disclose all the details, since it was for work.
